gems/smithy-schema/lib/smithy-schema/document_utils.rb

```diff
@@ -0,0 +1,231 @@
+          when Shapes::ListShape then apply_list(data, schema)
+          when Shapes::MapShape then apply_map(data, schema)
+          when Shapes::TimestampShape then apply_timestamp(data, schema)
+          when Shapes::BlobShape then Base64.decode64(data)
+          else data
+          end
+        end
+
+        # rubocop:disable Metrics/CyclomaticComplexity
+        def extract(data, schema, opts = {})
+          return if data.nil?
+
+          case shape(schema)
+          when Shapes::StructureShape then extract_structure(data, schema, opts)
+          when Shapes::UnionShape     then extract_union(data, schema, opts)
+          when Shapes::ListShape      then extract_list(data, schema)
+          when Shapes::MapShape       then extract_map(data, schema)
+          when Shapes::BlobShape      then extract_blob(data)
+          when Shapes::TimestampShape then extract_timestamp(data, schema, opts)
+          else data
+          end
+        end
+        # rubocop:enable Metrics/CyclomaticComplexity
+
+        private
+
+        def apply_list(data, schema)
+          shape = shape(schema)
+          data.map do |v|
+            next if v.nil?
+
+            apply(v, shape.member)
+          end
+        end
+
+        def apply_map(data, schema)
+          shape = shape(schema)
+          data.transform_values do |v|
+            if v.nil?
+              nil
+            else
+              apply(v, shape.value)
+            end
+          end
+        end
+
+        def apply_structure(data, schema, type)
+          shape = shape(schema)
+
+          type = shape.type.new if type.nil?
+          data.each do |k, v|
+            name =
+              if (member = member_with_json_name(k, shape))
+                shape.name_by_member_name(member.name)
+              else
+                member_name(shape, k)
+              end
+            next if name.nil?
+
+            type[name] = apply(v, shape.member(name))
+          end
+          type
+        end
+
+        def apply_timestamp(data, schema)
+          data = data.is_a?(Numeric) ? Time.at(data) : Time.parse(data)
+          time(data, timestamp_format(schema))
+        end
+
+        def apply_union(data, schema, type)
+          shape = shape(schema)
+          key, value = data.flatten
+          return if key.nil?
+
+          if (member = member_with_json_name(key, shape))
+            apply_union_member(member.name, value, shape, type)
+          elsif shape.name_by_member_name?(key)
+            apply_union_member(key, value, shape, type)
+          else
+            shape.member_type(:unknown).new(key, value)
+          end
+        end
+
+        def apply_union_member(key, value, shape, type)
+          member_name = shape.name_by_member_name(key)
+          type = shape.member_type(member_name) if type.nil?
+          type.new(apply(value, shape.member(member_name)))
+        end
+
+        def extract_blob(data)
```

Comment:
I know we're only supporting json right now, but blobs would be one of the things that would differ with other formats - eg: in a cbor document, we wouldn't base64 encode these.... I'm not sure how best to represent that behavior though and maybe its fine to leave in generic DocumentUtils for now since its api private and we could refactor it into JsonDocumentUtils, CborDocumentUtils, ect later if needed?